vms/platformvm/lock/camino_lock.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (C) 2022, Chain4Travel AG. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package lock
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/vms/components/avax"
+)
+
+var (
+	errInvalidLockState = errors.New("invalid lockState")
+	errNestedLocks      = errors.New("shouldn't nest locks")
+	ThisTxID            = ids.ID{'t', 'h', 'i', 's', ' ', 't', 'x', ' ', 'i', 'd'}
+)
+
+type LockState byte
+
+const (
+	LockStateUnlocked        LockState = 0b00
+	LockStateDeposited       LockState = 0b01
+	LockStateBonded          LockState = 0b10
+	LockStateDepositedBonded LockState = 0b11
+)
+
+var lockStateStrings = map[LockState]string{
+	LockStateUnlocked:        "unlocked",
+	LockStateDeposited:       "deposited",
+	LockStateBonded:          "bonded",
+	LockStateDepositedBonded: "depositedBonded",
+}
+
+func (ls LockState) String() string {
+	lockStateString, ok := lockStateStrings[ls]
+	if !ok {
+		return fmt.Sprintf("unknownLockState(%d)", ls)
+	}
+	return lockStateString
+}
+
+func (ls LockState) Verify() error {
+	if ls < LockStateUnlocked || LockStateDepositedBonded < ls {
+		return errInvalidLockState
+	}
+	return nil
+}
+
+func (ls LockState) IsBonded() bool {
+	return LockStateBonded&ls == LockStateBonded
+}
+
+func (ls LockState) IsDeposited() bool {
+	return LockStateDeposited&ls == LockStateDeposited
+}
+
+type LockIDs struct {
+	DepositTxID ids.ID `serialize:"true" json:"depositTxID"`
+	BondTxID    ids.ID `serialize:"true" json:"bondTxID"`
+}
+
+func (lock LockIDs) LockState() LockState {
+	lockState := LockStateUnlocked
+	if lock.DepositTxID != ids.Empty {
+		lockState = LockStateDeposited
+	}
+	if lock.BondTxID != ids.Empty {
+		lockState |= LockStateBonded
+	}
+	return lockState
+}
+
+func (lock LockIDs) Lock(lockState LockState) LockIDs {
+	if lockState.IsDeposited() {
+		lock.DepositTxID = ThisTxID
+	}
+	if lockState.IsBonded() {
+		lock.BondTxID = ThisTxID
+	}
+	return lock
+}
+
+func (lock LockIDs) Unlock(lockState LockState) LockIDs {
+	if lockState.IsDeposited() {
+		lock.DepositTxID = ids.Empty
+	}
+	if lockState.IsBonded() {
+		lock.BondTxID = ids.Empty
+	}
+	return lock
+}
+
+func (lock *LockIDs) FixLockID(txID ids.ID) {
+	if lock.DepositTxID == ThisTxID {
+		lock.DepositTxID = txID
+	}
+	if lock.BondTxID == ThisTxID {
```

Comment:
I would better use an else or return in case the first one matches.
Q: what is if both are already set? beside this, do you really want such an "insecure" fix?
It is dependent on the LockState, not really on the value of the tx

Reply:
> Q: what is if both are already set?

What do u mean by already set? If any of ids equal to ThisTxID, than it requires "fixing", cause outs can have ThisTxID only they are freshly locked in tx body - on execute UTXOs will be "fixed" and future outs produced from consumption of this UTXOs will have normal lockTxID.

> beside this, do you really want such an "insecure" fix?
It is dependent on the LockState, not really on the value of the tx

Yes, its purely lock.IDs method that just does ThisTxID replacement with txID. May be I should rename it to something like "ReplaceThixTxIDsWithRealID", dunno. This method is exported and used by package utxos ProduceLocked(), which is also exported (like utxos Produce()) and used in tx executions